### PR TITLE
Improve waveform classes & derivatives numerical precision

### PR DESCRIPTION
**Pull request by AlessandroAgapito23**

### CODE REVIEW

File: GWFish/modules/waveforms.py [hunk 1]
```diff
@@ -519,51 +520,59 @@ def calculate_frequency_domain_strain(self):
...
+        eta2 = eta*eta
+        eta3 = eta2*eta
+
+        chi_eff = (M1*chi_1 + M2*chi_2)/M
+        chi_PN = chi_eff - 38/113*eta*(chi_1 + chi_2)
+        chi_s = 0.5*(chi_1 + chi_2)
+        chi_a = 0.5*(chi_1 - chi_2)
+
+        #f_cut = cut_order * f_isco
+        cut = self.gw_params['cut']
```
Comment (jacopok) [inline GWFish/modules/waveforms.py:539]: Is this something that needs to be given by the user? I think we should at least a default (e.g. `cut = self.gw_params.get('cut', 4)`)

File: GWFish/modules/inspiral_corrections.py [hunk 2]
```diff
@@ -0,0 +1,912 @@
...
+    #t(f)=(1/2pi)*phi_prime(f) at lowest order in stationary phase approximation
+    t_of_f = -5./(256.*np.pi**(8/3))/Mc**(5/3)/frequencyvector**(8/3)
+
+    return t_of_f+parameters['geocent_time']
+
+################################################################################
+################### WAVEFORMS PARAMETERS AND VARIABLES #########################
+################################################################################
+
+class Waveform:
```
Comment (jacopok) [inline GWFish/modules/inspiral_corrections.py:74]: Is this a redefinition of the Waveform class from the `waveforms.py` module? If so, that's dangerous... If you need to modify its functionality you can use inheritance (or we can discuss and change the original one)

File: GWFish/modules/inspiral_corrections.py [hunk 3]
```diff
@@ -0,0 +1,912 @@
...
+import logging
+import requests 
+from io import StringIO
+import matplotlib.pyplot as plt
+import numpy as np
+import sympy as sp
+from scipy.interpolate import interp1d
+import scipy.optimize as optimize
+
+#LAL libraries
```
Comment (jacopok) [inline GWFish/modules/inspiral_corrections.py:13]: Do you need lal in this module? It seems like it's not used anywhere....

File: GWFish/modules/fishermatrix.py [hunk 4]
```diff
@@ -234,7 +243,7 @@ def compute_detector_fisher(
     signal_parameter_values: Union[pd.DataFrame, dict[str, float]],
     fisher_parameters: Optional[list[str]] = None,
     waveform_model: str = wf.DEFAULT_WAVEFORM_MODEL,
-    waveform_class: type(wf.Waveform) = wf.LALFD_Waveform,
+    waveform_class = wf.LALFD_Waveform,
```
Comment (jacopok) [inline GWFish/modules/fishermatrix.py:246]: Why is the type hint removed here?
Eventually we should get rid of the `waveform_class` parameter altogether so it doesn't matter that much, but I'm curious: was it giving you trouble?

File: GWFish/modules/fishermatrix.py [hunk 5]
```diff
@@ -46,6 +46,11 @@ def fft_derivs_at_detectors(deriv_list, frequency_vector):
     idx_f_high = int(frequency_vector[-1,0]/delta_f)
 
     return np.vstack(ffd_deriv_list).T[idx_f_low:idx_f_high+1,:]
+
```
Comment (jacopok) [inline GWFish/modules/fishermatrix.py:49]: Not a huge fan of the headers to be honest but that's just personal preference

**Comment by jacopok** [comment]
Thanks for this @AlessandroAgapito23 ! I think the changes are overall good, but there are some rough edges to take care of. 
See my comments above. 
Also, you can see that the automated tests are failing - some of them were using GWFish's `TaylorF2`, and something is going wrong there with the spin variables. You can see the details from the Github interface here, or you can rerun the tests locally (see [development installation](https://gwfish.readthedocs.io/en/latest/installation.html#development-installation)).

If you want (not compulsory for this to be merged), it would also be great to have some tests for the new PPE waveforms.
These could go into the [`test/test_waveforms.py`](https://github.com/janosch314/GWFish/blob/main/tests/test_waveforms.py) file, or another one as you prefer. 
For example, a good and simple one might be to check that the PPE waveforms reduce to the regular TaylorF2 when the PE corrections are zero.
You can see what the basic syntax of the tests looks like by looking at the ones that are already present; also you can check out the [pytest docs](https://docs.pytest.org/en/8.0.x/).

Finally, since the waveforms module was originally @bvgoncharov 's child I think we could benefit from his opinion as well.

File: GWFish/modules/inspiral_corrections.py (hunk 2 above)
Comment (AlessandroAgapito23) [reply GWFish/modules/inspiral_corrections.py:74]: No, there is no redefinition of the Waveform class. The only modification is regarded to the addition of the ppE parameters and f_cut in self.gw_params. Maybe I could modify this directly in waveforms.py

File: GWFish/modules/fishermatrix.py (hunk 4 above)
Comment (AlessandroAgapito23) [reply GWFish/modules/fishermatrix.py:246]: No trouble at all. I wanted to see "compute_detector_fisher" and "compute_network_errors" with the same input syntax.

File: GWFish/modules/waveforms.py (hunk 1 above)
Comment (jacopok) [reply GWFish/modules/waveforms.py:552]: Not what I meant, this still requires the user to go through the code and read the comment - I mean a default given in a way that one can _not specify_ `'cut'` as part of the `params` dictionary.

Comment (AlessandroAgapito23) [reply GWFish/modules/waveforms.py:552]: I don't understand, sorry. The default value (4) is in gw_params and, in principle, one can not specify 'cut' as part of the parameters without having trouble.

Comment (jacopok) [reply GWFish/modules/waveforms.py:552]: Ah! missed it, you are right.

File: GWFish/modules/inspiral_corrections.py [hunk 6]
```diff
@@ -0,0 +1,795 @@
...
+        
+        hp = amp_tot*0.5*(1 + np.cos(iota)**2.)
+        hc = amp_tot*np.cos(iota)
+        polarizations = np.hstack((hp * phase, hc * 1.j * phase))
+
+        ############################### OUTPUT #################################
+
+        self._frequency_domain_strain = polarizations
+
+        ########################################################################
```
Comment (bvgoncharov) [inline GWFish/modules/inspiral_corrections.py:792]: This also copies over most of the code from TaylorF2 and PhenomD, we don't do that in Python :) 

Instead, we should break the standard code `waveforms.py` into more functions and attributes (self.something). Then, for PPE corrections, we should make child classes here in `inspired_corrections.py`, and PPE corrections should be added by overwriting specific attributes/functions in a short way, such that we don't copy the code over. Can you do this, @AlessandroAgapito23 ?

We can discuss it on the next Thursday's meeting from 2pm to 4 pm, and perhaps make a quick code session together on Zoom (CC @jacopok, @u-dupletsa , @janosch314  ).

File: GWFish/modules/waveforms.py [hunk 7]
```diff
@@ -159,7 +159,20 @@ def _set_default_gw_params(self):
...
+            'beta':0., 'PN':0.,
+            #gIMR
+            'delta_phi_0':0.,
+            'delta_phi_1':0.,
+            'delta_phi_2':0.,
+            'delta_phi_3':0.,
+            'delta_phi_4':0.,
+            'delta_phi_5':0.,
+            'delta_phi_6':0.,
+            'delta_phi_7':0.,
```
Comment (bvgoncharov) [inline GWFish/modules/waveforms.py:173]: These are PPE-specific parameters, I recommend to make a child class of the LAL waveform in `inspiral_corrections.py` with these parameters, and remove them from here.

File: GWFish/modules/waveforms.py (hunk 1 above)
Comment (bvgoncharov) [reply GWFish/modules/waveforms.py:552]: @AlessandroAgapito23 , around lines ≈507-590, you seemed to change `class TaylorF2` quite a lot. I do not understand what you did from looking at the changes, the code pieces that are added and removed do not correspond to each other. I would not change our standard TaylorF2 for the sake of reproducibility, so perhaps we should revert back to the old code. 

As I mentioned in the earlier comment, I would then break TaylorF2 and PhenomD into more functions and attributes, such that in `inspiral_corrections.py` we do not copy the code over from `waveforms.py`, instead we should only add new lines of code with PPE corrections.

**Comment by bvgoncharov** [comment]
Hi @AlessandroAgapito23 , @jacopok , I reviewed the code, please see my comments above

**Comment by AlessandroAgapito23** [comment]
Hi @bvgoncharov, unfortunately I don't see any comments above. Where can I find them?

**Review by bvgoncharov — CHANGES_REQUESTED** [review]
Ok, hopefully my comments will appear now, after this message

File: GWFish/modules/fishermatrix.py [hunk 8]
```diff
@@ -55,7 +60,7 @@ class Derivative:
 
     eps: 1e-5, this follows the simple "cube root of numerical precision" recommendation, which is 1e-16 for double
     """
-    def __init__(self, waveform, parameters, detector, eps=1e-5, waveform_class=wf.Waveform):
+    def __init__(self, waveform, parameters, detector, eps=1e-9, waveform_class=wf.Waveform):
```
[the remark this answers is not in the document]
Comment (AlessandroAgapito23) [reply GWFish/modules/fishermatrix.py:63]: I forgot to change eps = 1e-9 to eps = 1e-5, now I'm going to leave eps = 1e-5.
Thanks a lot for the comment, I needed this numerical precision for one of my analyses.
I can call the class "fishermatrix (eps = 1e-9)" directly from my py script.

File: GWFish/modules/inspiral_corrections.py [hunk 9]
```diff
@@ -0,0 +1,795 @@
...
+    M2 = local_params['mass_2'] * cst.Msol
+
+    M = M1 + M2 #Total mass
+    mu = M1 * M2 / M #Reduced mass
+    Mc = cst.G * mu ** 0.6 * M ** 0.4 / cst.c ** 3 # Chirp mass in seconds
+
+    #t(f)=(1/2pi)*phi_prime(f) at lowest order in stationary phase approximation
+    t_of_f = -5./(256.*np.pi**(8/3))/Mc**(5/3)/frequencyvector**(8/3)
+
+    return t_of_f+parameters['geocent_time']
```
[the remark this answers is not in the document]
Comment (AlessandroAgapito23) [reply GWFish/modules/inspiral_corrections.py:58]: Sure, I'm going to cut them from the module.
Moreover, I think they aren't useful in inspiral_corrections.py.

**Comment by AlessandroAgapito23** [comment]
Between line 507 and 590, so regarding "calculate_frequency_domain_strain() function", I changed the following things :

- Definition of the frequency as "frequencyvector" instead of "ff" which is a dimensionless frequency in IMRPhenomD
- Addition of the spin-corrections into the phi_i terms of the psi_TF2 phase, as in IMRPhenomD and, by doing that, I have used the same syntax between the two models
- Addition of a different f_cut (respect to f_cut = 4 * f_isco) in the gw_parms ( default 'cut' : 4) 

Note : I was wrong in writing the psi_tot in "TaylorF2", there was a typo in the self.psi definition.
Now I think that the "TaylorF2" waveform should work, I'm going to check it.

File: GWFish/modules/fishermatrix.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (AlessandroAgapito23) [reply GWFish/modules/fishermatrix.py]: I am sorry about that, maybe I should have been more precise about this point.
One of the main goals, of my ongoing ppE analysis, was to reproduce and compare (with fig.4 of arXiv:1603.08955)  how these so-called "ppE inspiral corrections" constraints vary as functions of the PN order (roughly a linear trend between 10-8 and 10+3).
Briefly, two types of constraints disagreements, respectively at two different PN regimes, sorted out with respect to arXiv:1603.08955:

- under-estimation at PN > 1.5
- plateau at PN < -2

I realized that the first one was due to the fact that in GWFish the derivatives of the phase terms are taken numerically. Writing down the analytical derivatives of the phase terms, I obtained a very good agreement at PN > 1.5.
After a brief discussion with Jacopo, about the second type of disagreement, I tried to change eps to lower values such as 1e-6, 1e-7, and 1e-8. The constraints plateau disappeared and I was able to obtain the minimum constraint at PN = - 4 which is almost 10-8 (same as in arXiv:1603.08955).

File: GWFish/modules/inspiral_corrections.py (hunk 6 above)
Comment (AlessandroAgapito23) [reply GWFish/modules/inspiral_corrections.py]: Hi all, I have worked on this type of modification in waveforms.py and inspiral_corrections.py.
I have also added a brief comment at the beginning of the script, in order to explain and resume all the new functions and attributes.
I was also thinking, upon Jacopo's suggestion, about some tests on the new waveforms. I'm going to work on them.

**Pull request closed without merging**